### PR TITLE
Fzchriha/labgraph UI 

### LINKED ISSUE #45: LabGraph Monitor Improvement - Add input/output message information in LabGraph Monitor UI
## 🚀 Feature
Add input/output message information (e.g. data type and message name) in LabGraph Monitor GUI (frontend)

## Additional context

1. Existing application can be found [here] (https://github.com/facebookresearch/labgraph/pull/38)
2. The code should be added at folder is https://github.com/facebookresearch/labgraph/pull/38
3. Create setup.py and README.md, where example can be found [here](https://github.com/facebookresearch/labgraph/tree/main/extensions/labgraph_viz)
4. Add github action support, reference: [here](https://github.com/facebookresearch/labgraph/actions/workflows/main.yml)
5. Add proper license header, [example](https://github.com/facebook/react/blob/main/scripts/babel/transform-object-assign-require.js).

### PR DESCRIPTION
**Pull request by fzchriha**
## Description

Create UI for real-time messaging and set up the logic behind real-time data rendering.

Fixes #45 

## Demo
demo_realtimedata2

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [X] New and existing unit tests pass locally with these changes?
- [X] Has code been commented, particularly in hard-to-understand areas?
- [X] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Comment by fzchriha** [comment]
@bennaaym I added the test for the existence of a .env.local file
Here is how it is displayed
labgraph_handle_error

Please let me know if you have any other suggestions?

**Comment by bennaaym** [comment]
@fzchriha, maybe the message could be updated to ````Undefined Environment Variable: REACT_APP_WS_API````, just to be more specific. otherwise looks good. great work!

**Comment by fzchriha** [comment]
@jfResearchEng Hi Jimmy here is my final PR, I have made all the requested changes as well as added additional UI features & tests. Can you please review my code?

File: extensions/prototypes/labgraph_monitor/src/components/SettingPanel/EdgeSettings.tsx [hunk 1]
```diff
@@ -62,20 +93,64 @@ const Edge: React.FC = (): JSX.Element => {
...
+                                                            style={{
+                                                                whiteSpace:
+                                                                    'normal',
+                                                                wordBreak:
+                                                                    'break-word',
+                                                            }}
+                                                        >
+                                                            {mockData.join(' ')}
+                                                        </TableCell>
+                                                    )}
```
Comment (navn-r) [inline extensions/prototypes/labgraph_monitor/src/components/SettingPanel/EdgeSettings.tsx:149]: This nested ternary is a bit hard to follow
```suggestion
                                                    {open ? (
                                                        <TableCell
                                                            style={{
                                                                whiteSpace:
                                                                    'normal',
                                                                wordBreak:
                                                                    'break-word',
                                                            }}
                                                        >
                                                            {connection === WS_STATE.CONNECTED ? 
                                                                 `${field[1].content}, ` : mockData.join(' ')
                                                             }
                                                        </TableCell>
                                                    ) : null}
```

**Review by navn-r — COMMENTED** [review]
Overall, great work! :tada: 🌮 

Just left some nitpicks and code suggestions 🚀. Apologies if the suggestions formatting is a bit off...the text editor on GitHub for suggestions sucks.

**Comment by jfResearchEng** [comment]
Just to confirm with you, whether this PR can be used for final review and the rest of your LabGraph UI drafts can be ignored for now?

File: extensions/prototypes/labgraph_monitor/src/redux/reducers/graph/mock/interfaces/IMock.ts [hunk 2]
```diff
@@ -8,6 +8,7 @@ import IGraph from '../../common/interfaces/IGraph';
 
 interface IMock {
     mockGraph: IGraph;
+    mockRealtimeData: any;
```
[the remark this answers is not in the document]
Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/src/redux/reducers/graph/mock/interfaces/IMock.ts:11]: Hi Nav since the type of mockRealTimeData is a combination of types, I wanted to make life easy so I used *any*

File: extensions/prototypes/labgraph_monitor/src/components/SettingPanel/EdgeSettings.tsx [hunk 3]
```diff
@@ -14,15 +14,22 @@ import {
...
 import WS_STATE from '../../redux/reducers/graph/ws/enums/WS_STATE';
+import { setMockRealtimeData } from '../../redux/reducers/graph/mock/mockReducer';
 
 interface IMessage {
     name: string;
-    fields: { [fieldName: string]: string };
+    fields: {
+        [fieldName: string]: {
+            type: string;
+            content: any;
```
[the remark this answers is not in the document]
Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/src/components/SettingPanel/EdgeSettings.tsx:30]: Thanks, Nav for the great feedback, unfortunately, content can be either a float or an Array. That's why it was hard to specify

`"Attenuator": [{
        "name": "RandomMessage",
        "fields": {
            "timestamp": {
                "type": "float",
                "content": 1647047794.474803,
            },
            "data": {
                "type": "ndarray",
                "content": array([0.77135353, 0.61299748, 0.34146919, 0.46154968, 0.19577749],
            }
        }
    }
]`

File: extensions/prototypes/labgraph_monitor/src/components/SettingPanel/SettingPanel.tsx [hunk 4]
```diff
@@ -106,6 +149,20 @@ const SettingPanel: React.FC = (): JSX.Element => {
...
+                        onMouseDown={(e) => handleMouseDown(e)}
+                        className={classes.dragger_light}
+                    />
+                ) : (
+                    <div
+                        id="dragger"
+                        onMouseDown={(e) => handleMouseDown(e)}
+                        className={classes.dragger_dark}
+                    />
+                )}
```
[the remark this answers is not in the document]
Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/src/components/SettingPanel/SettingPanel.tsx:165]: `Compiled with problems:X

ERROR in src/components/SettingPanel/SettingPanel.tsx:156:32

TS7053: Element implicitly has an 'any' type because expression of type 'string' can't be used to index type 'ClassNameMap<"root" | "settingButton" | "settingIcon" | "settingPanel" | "themeBar" | "dragger_light" | "dragger_dark">'.
  No index signature with a parameter of type 'string' was found on type 'ClassNameMap<"root" | "settingButton" | "settingIcon" | "settingPanel" | "themeBar" | "dragger_light" | "dragger_dark">'.
    154 |                     id="dragger"
    155 |                     onMouseDown={handleMouseDown}
> 156 |                     className={classes['dragger_' + mode]}
        |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^
    157 |                 />`
    
    I get this error when I try the above suggestion

[the remark this answers is not in the document]
Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/src/components/SettingPanel/SettingPanel.tsx:165]: Used this suggestion instead
`className={mode === 'light' ? classes.dragger_light : classes.dragger_dark}`

**Comment by jfResearchEng** [comment]
Has this PR been rebased and included the latest changes made in PR #64 ?

**Comment by fzchriha** [comment]
@jfResearchEng Yes I tested my changes with Damir's latest changed

File: extensions/prototypes/labgraph_monitor/src/redux/reducers/graph/mock/interfaces/IMock.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (navn-r) [reply extensions/prototypes/labgraph_monitor/src/redux/reducers/graph/mock/interfaces/IMock.ts:11]: Oh I see, well TypeScript has first-class support for unioning types. Look into https://www.typescriptlang.org/docs/handbook/utility-types.html you might find it useful 👍

File: extensions/prototypes/labgraph_monitor/README.md [hunk 5]
```diff
@@ -71,7 +74,7 @@ REACT_APP_WS_API="ws://127.0.0.1:9000"
 
 (!) LabGraph Websocket server runs on localhost:9000 by default
 
-2. Run Labgraph Websockets server. The following tutorial shows how to run LabGraph Websocket server properly : [tutorial](https://github.com/facebookresearch/labgraph/pull/58/files#diff-247005c77570899ce53f81a83b2a5fe6e7535616cc96564d67378fe7f73dac49)
+2. Run Labgraph Websockets server. The following tutorial show how to run LabGraph Websocket server properly : [tutorial](https://github.com/facebookresearch/labgraph/pull/58/files#diff-247005c77570899ce53f81a83b2a5fe6e7535616cc96564d67378fe7f73dac49)
```
Comment (jfResearchEng) [inline extensions/prototypes/labgraph_monitor/README.md:77]: A merged version link is preferred over a PR link.

Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/README.md:77]: Yes I will update that

File: extensions/prototypes/labgraph_monitor/src/redux/reducers/graph/common/interfaces/INode.ts [hunk 6]
```diff
@@ -10,7 +10,10 @@ interface INode {
             [upstream: string]: Array<{
                 name: string;
                 fields: {
-                    [fieldName: string]: string;
+                    [fieldName: string]: {
+                        type: string;
+                        content: any;
```
[the remark this answers is not in the document]
Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/src/redux/reducers/graph/common/interfaces/INode.ts:15]: resolved

File: extensions/prototypes/labgraph_monitor/src/components/SettingPanel/SettingPanel.tsx [hunk 7]
```diff
@@ -68,7 +94,23 @@ const SettingPanel: React.FC = (): JSX.Element => {
...
     const { panelOpen } = useSelector((state: RootState) => state.config);
     const dispatch = useDispatch();
+    const [panelWidth, setPanelWidth] = React.useState(defaultPanelWidth);
 
+    const handleMouseDown = (e: any) => {
+        document.addEventListener('mouseup', handleMouseUp, true);
+        document.addEventListener('mousemove', handleMouseMove, true);
+    };
+    const handleMouseUp = () => {
+        document.removeEventListener('mouseup', handleMouseUp, true);
```
[the remark this answers is not in the document]
Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/src/components/SettingPanel/SettingPanel.tsx:104]: Hmm I didn't understand, I used a sample code from elsewhere I tweaked it a bit

**Comment by fzchriha** [comment]
@jfResearchEng I updated the README file with the latest changes from PR64

File: extensions/prototypes/labgraph_monitor/README.md [hunk 8]
```diff
@@ -21,28 +21,31 @@ yarn -v
...
 ```
-yarn
+labgraph\extensions\prototypes\labgraph_monitor> yarn
 ```
 
 3. Test the application by running the following command
 
 ```
-yarn test --watchAll=false
+labgraph\extensions\prototypes\labgraph_monitor> yarn test --watchAll=false
```
Comment (jfResearchEng) [inline extensions/prototypes/labgraph_monitor/README.md:42]: The following error appears when running this command: https://paste.ofcode.org/VFn3XgZGCsnF3hiTGKNXuZ

File: extensions/prototypes/labgraph_monitor/README.md [hunk 9]
```diff
@@ -21,28 +21,31 @@ yarn -v
...
 ```
 
 4. Run the application
 
 ```
-yarn start
+labgraph\extensions\prototypes\labgraph_monitor> yarn start
```
Comment (jfResearchEng) [inline extensions/prototypes/labgraph_monitor/README.md:48]: An empty UI has launched after running this line.
Reference: https://pasteboard.co/K6noFCNDD7A3.png

Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/README.md:48]: Do you the icon on the right?
Screen Shot 2022-04-22 at 2 33 13 PM

File: extensions/prototypes/labgraph_monitor/README.md (hunk 8 above)
Comment (fzchriha) [reply extensions/prototypes/labgraph_monitor/README.md:42]: I resolved it, thank you for pointing that out

**Review by jfResearchEng — APPROVED** [review]
Merging this PR for now and for the other improvements, please help create a new PR.

**Comment by jfResearchEng** [comment]
Thank you for contributing to all the PR improvements on LabGraph Monitor!
https://c.tenor.com/as2lvRfV6nUAAAAd/dolunay-bittersweet.gif

**Comment by jfResearchEng** [comment]
@fzchriha, is this still an open issue: https://github.com/facebookresearch/labgraph/issues/69 ?
